### ISSUE TITLE
[Bug] ttnn.log2 is not exact for powers of two in float32 (46 of 254 exponents off by an ULP)

### ISSUE THREAD

**Issue by AJ0070** [tenstorrent/tt-metal#52665]
### Component / Area

ML operations (SFPU / LLK), `ttnn.log2`

### Issue Type (optional)

Bad Outputs

### Observed

`ttnn.log2` does not return the exact exponent for an exact power of two in float32.
**46 of the 254 representable exponents come back an ULP low.**

Measured on an N300 (Wormhole B0), sweeping `2**k` for k in [-126, 127]:

```
powers of two returning the exact exponent: 208/254

log2(2**-125)  device -124.99999237   expected -125
log2(2**-124)  device -123.99999237   expected -124
log2(2**-120)  device -119.99999237   expected -120
log2(2**-119)  device -118.99999237   expected -119
```

**Root cause.** There is no dedicated log2 kernel. `ttnn.log2` is dispatched as natural
log multiplied by a base constant
(`ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp`,
`log_with_base_tile(idst, 0x3fb8aa3bu)` where `0x3fb8aa3b` is 1/ln(2)). In
`ckernel_sfpu_log.h` the natural log is formed first and the base change is applied to
the finished sum:

```cpp
result = e_float * sfpi::vConstFloatPrgm0 + r;   // vConstFloatPrgm0 = ln(2) * 2^-23

if constexpr (HAS_BASE_SCALING) {
    result *= sfpi::as<sfpi::vFloat>(sfpi::vUInt(log_base_scale_factor));
}
```

That multiply also scales the exponent contribution, so the exponent path effectively
gets multiplied by `ln(2) * (1/ln(2))`. Those are two separately rounded float
constants and their product does not round to exactly 1, so the exponent term is
perturbed. For an exact power of two the mantissa term `r` is 0 and the entire result
is that perturbed exponent, which is why the error is visible precisely at the inputs
where the answer should be exact.

The mantissa term does not have this problem, because scaling it is the mathematically
correct operation. Only the exponent term needs to stay unscaled.

Note bfloat16 is unaffected (254/254 exact): the error is below bf16 resolution.

### Expected

`log2(2**k) == k` exactly for every representable exponent, as it is on the CPU
reference and in bfloat16 on device today.

### 1. Steps (exact commands)

```python
import torch
import ttnn

dev = ttnn.open_device(device_id=0)
ks = list(range(-126, 128))
buf = torch.ones((1, 1, 32, 256), dtype=torch.float32)
buf[0, 0, 0, : len(ks)] = torch.tensor([2.0**k for k in ks], dtype=torch.float32)

t = ttnn.from_torch(buf, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=dev)
got = ttnn.to_torch(ttnn.log2(t))[0, 0, 0, : len(ks)].float()

bad = [(ks[i], float(got[i])) for i in range(len(ks)) if float(got[i]) != float(ks[i])]
print("exact:", len(ks) - len(bad), "/", len(ks))
print("first few wrong:", bad[:4])
ttnn.close_device(dev)
```

### 2. Input data / link or description

No input data required. Any exact power of two in the affected set reproduces it.

### 3. Frequency

Consistent and deterministic.

### 1. Software Versions

tt-metal main at 003feb6663.

### 2. Hardware Details

Wormhole B0, N300. The same code path is present in the `blackhole` copy of
`ckernel_sfpu_log.h`.

### Is this a regression?

Unknown. The base-scaling multiply has been applied to the finished sum for as long as
the current log kernel has existed.

### Logs & Diagnostics

Applying the base change to the mantissa term only, for base 2, makes the remaining
exponent factor exactly `2^-23`. Since the exponent arrives as `k << 23` and `2^-23` is
a power of two, the product is exact. Measured on the same N300 after that change:

| metric (fp32) | before | after |
| --- | --- | --- |
| powers of two exact | 208/254 | **254/254** |
| max relative error vs float64 | 1.026e-07 | **6.764e-08** |
| mean relative error vs float64 | 3.080e-08 | **2.370e-08** |

`log10` and plain `log` are unaffected by that change (every metric identical before and
after), as are all special values. Instruction count is unchanged, and on a 4096x4096
fp32 tensor `log2` measures 0.8505 ms/call against 0.8576 ms/call for `log10`, which
still uses the old base-scaled path in the same build.

Worth noting for anyone tempted to generalise it: the equivalent base 10 constant
`log10(2) * 2^-23` is not exactly representable, and folding it in measured **worse**
than the existing path (powers of ten exact dropped from 59/76 to 47/76). The fix
should stay scoped to base 2.

A fix is proposed in the linked PR.

### Priority

Low

### Impact

Correctness and IEEE conformance rather than a large numerical error: the existing worst
case is about 1 ULP. It matters for code that relies on `log2` of a power of two being
exact, for example deriving an integer shift or bit width, where an answer of
-124.99999237 truncates to a different integer than -125.

**Comment by AJ0070** [comment]
@cmaryanTT Can you please trigger CI on the attached PR?
